RIOT-OS/RIOT
# build system: add list of features for documentation and sanity checking

### Contribution description

This PR adds a `features.yaml` file as machine readable but easy to edit format that lists, groups, and documents all features ("feature" as in the the build system entity) in RIOT. A small python script generates a Makefile for and markdown file from that.

The Makefile is used to check if all features provided and requested are actually valid names or typos.

The Markdown file is used by Doxygen to generation a documentation page containing all features.

### Testing procedure

1. Try requesting or providing a non-existing feature. A build for an app+board combo that contains a feature provided or requested that does not exist should fail.
    - Beware: This requires `make generate-features` to be run after touching `features.yaml`, as the generated Makefile is tracked as source. This avoids running a python script before each and every of the current 141584 build targets (which would result in 4h of additional CPU time per CI build, if each build would add 0.1s of CPU time.)
2. Check the generated documentation.
    - Here the markdown for consumption by Doxygen is generated as part of `make doc`, so this would work right away
3. Check that `make static-tests` catches when the list of features in `makefiles/features_existing.inc.mk` is stale (e.g. you updated `features.yaml` but haven't run `make generate-features` yet).

### Issues/PRs references

- [x] Requires https://github.com/RIOT-OS/riotdocker/pull/247
